### DISCUSSION TITLE
I2S confusion support for controller and/or peripheral operation

### DISCUSSION THREAD

**Discussion by saraverbeecke** [micropython/micropython#11675] [ESP32] [answered]
Hi,
the docs state: `The I2S class supports controller operation. Peripheral operation is not supported.` Does this mean that only I2S devices in Master mode are supported (= I2S device generates its own system clock)?

I'm trying to get audio in on an ESP32S3 (ESP-IDF v4.4) through a [PCM1808](https://www.ti.com/lit/ds/symlink/pcm1808.pdf?ts=1685513731156) ADC:

1. should this device be set as a master or slave?
2. should I generate the SCKI with a PWM from the ESP32 at 256*fs (seems to be necessary for both slave or master mode, PWM solution was suggested in the old MicroPython forum, is this reliable enough?)
3. 24-bit data should come out over I2S, is there any way to process the buffer fast containing the 32 bit values (only 32 bit or 16 bit mode is support in the library and it is all little endian I suppose) but needs to be converted to integers for further processing...
4. how do I best separate the L and R channel values?
5. I took a look at the great @miketeachman examples, but still could not figure it out for this ADC.

```
from machine import Pin, I2S, PWM
from struct import unpack

BUFFER_LENGTH_IN_BYTES = 4096
USER_BUFFER_LENGTH_IN_BYTES = round(BUFFER_LENGTH_IN_BYTES / 4)
RATE = 16_000

pwm = PWM(Pin(1), freq=256*RATE, duty=512) # is this ok ???
i2s = I2S(0, sck=Pin(2), sd=Pin(42), ws=Pin(41), mode=I2S.RX, bits=32, format=I2S.STEREO, rate=RATE, ibuf=BUFFER_LENGTH_IN_BYTES)

samples = memoryview(bytearray(USER_BUFFER_LENGTH_IN_BYTES))

while True:
    num_bytes_read = i2s.readinto(samples) # read samples from the I2S peripheral
    if num_bytes_read > 0:
        print(samples[0:8])
        print(struct.unpack("<ii", samples[0:8])) # trying to understand
```

Thanks!

**Comment by miketeachman** [comment, marked as answer]
I took a quick look at the datasheet and it seems like this device should work with the machine.I2S class.

1. The ADC device should be configured in slave mode.  
2. Yes.  Use machine.PWM to create a 256*fs clock.

To extract 32-bit samples for L/R you'll need to iterate over the data that is read into the bytearray.  Here is some **untested** code that might help.  You'll need to add some additional code to test a condition to exit the while loop.  The real micropython experts will likely have a better way to do this, but this is a start.

``` 
left_samples = []
right_samples = []
while True:
    num_bytes_read = i2s.readinto(samples) # read samples from the I2S peripheral
    for i in range(0, len(samples ), 8):
        left_sample = struct.unpack('<i', samples [i:i+4])[0]  
        left_samples.append(left_sample)
        right_sample = struct.unpack('<i', samples [i+4:i+8])[0]
        right_samples.append(right_sample)
```
Does this help?

**Reply by saraverbeecke** [reply]
@miketeachman thanks for the insight that helped a bit but still cannot get a decent data out of the PCM1808. The code you suggest could indeed work (I would also ignore the first byte as the device only supports 24 bit so the first byte contains \x00).
I put the PCM1808 in slave mode and in I2S format (FMT, MD0, MD1 pins connected to ground with a 10k). Also added the suggested decoupling caps on the VCC, VDD and Vref rails. Then I applied the PWM at to create a 256*fs clock. The PWM frequency is not exact 256*fs as you know. Using an exact PWM frequency of 256*fs also does not help.

I always get full scale, zero or random data on the L/R channels...

Anything you can think of I forgot? 
Is it an option to set the device in master mode with a separate clock circuit with crystal and use the SCK and WS as outputs?

**Comment by miketeachman** [comment]
Sorry, not having used this device I'm struggling to offer other insights.  Do you have access to an oscilloscope to verify the various inputs, like master clock, sck, and ws?  I often consult the [schematic of the evaluation board](https://www.ti.com/lit/ug/sbau349/sbau349.pdf) for ideas.  Or, potentially [purchase the evaluation board](https://www.digikey.com/en/products/detail/texas-instruments/PCM1808EVM/15856747?s=N4IgTCBcDaIA4GMC2BGAHABjSAugXyA)?

**Reply by saraverbeecke** [reply]
@miketeachman sorry for the late reply. I got the evaluation board and have some values coming in on the I2S but the recorded signal seems to be distorted. I have a signal generator producing a 1 kHz sine wave at 1 Vpp on both L and R inputs but when plotting the digitized signal it is distorted heavily. Could this be due to the fact that I create the clock signal (PWM from the ESP32S3) for the slave device that might not be stable (i.e. jitter...)? The rise and fall times seems to be in spec with the datasheet.

1. Is there no way to use the device as master using an external clock source using MicroPython I2S?
2. If I use a PWM as MCK for the device, will it by in "sync" with the WS and SCK clock outputs of the ESP32S3? I was wondering this since the I2S for the mimxrt port has this feature. Posts on the Texas Instruments forum point out the fact that all I2S clocks can have a small phase difference, but this phase difference must be fixed in time so no jitter relative to each other must be present.

The clocks on my scope:

CH1 (yellow) = SCKI
CH2 (blue) = BCK
CH3 (magenta) = LRCK

![CH1_SCKI_CH2_BCK](https://github.com/micropython/micropython/assets/113063712/48e74bf6-e6b5-4609-aa76-5566b042ff90)
![CH3_LRCK](https://github.com/micropython/micropython/assets/113063712/91b10020-0043-4c2e-972a-48915bf47823)

**Reply by saraverbeecke** [reply]
I had some noise on the GND that made the clock jitter too much. I can now use a PWM as clock for the ADC. So I can confirm this method works.

**Reply by miketeachman** [reply]
WooHoo !  That's great news.  Thanks for preserving on this issue.

**Reply by frohro** [reply]
@saraverbeecke, I just found this, and I am delighted.  @miketeachman gave me a good bit of help too.  I'm still working on it.  It seems a bit amazing that you can generate the master clock independently of the I2S library with PWM and still have everything still work.  The datasheet says in Section 7.3.3, "In slave mode, the PCM1808 device operates under LRCK (pin 7), synchronized with system clock SCKI (pin 6).
The PCM1808 device does not require a specific phase relationship between LRCK and SCKI, but does require
the synchronization of LRCK and SCKI."  I am wondering if what you got working is a little by coincidence, because I don't see how the SCKI signal you generated with PWM is guaranteed to be "synchronized".   I think the PCM1808 in master mode would take care of the synchronization, but the [source](https://github.com/micropython/micropython/blob/master/ports/rp2/machine_i2s.c) looks these pins are outputs, so the only way micropython will work is as Master.  To me, this needs to be spelled out in the docs.   I would also be a lot more comfortable if I understood why a PWM-generated clock will remain in sync with the BCK and LRCK.  It may be because I don't understand exactly how the PWM generates a clock of arbitrary frequency.  In fact, if I ask PWM to generate me a 256*48000=1228800 clock, I get a 12500000 frequency according to pwm0.freq(), and that is a long way from what I wanted (48 kHz).  Is our sample rate all off when we generate the master clock from PWM?
Also, if you have any code or other things you would be willing to share with me I would be delighted!
Thanks to both of you!

**Reply by frohro** [reply]
@saraverbeecke I wish your scope photos showed the frequency measurement of the master clock.  :-)

**Reply by frohro** [reply]
@saraverbeecke and @miketeachman:

So I soldered up my PCM1808 Module to the Pico W, got out my oscilloscope, put a 2 kHz sine wave into the module offset appropriately.  I got only zeros.  Here is the code.  It is essentially your code above.

`from machine import Pin, I2S, PWM
import struct

BUFFER_LENGTH_IN_BYTES = 4096
USER_BUFFER_LENGTH_IN_BYTES = round(BUFFER_LENGTH_IN_BYTES / 4)
RATE = 48000

pwm0 = PWM(Pin(3), freq=256*RATE, duty_u16=32768)
i2s = I2S(0, sck=Pin(0), sd=Pin(2), ws=Pin(1), mode=I2S.RX, bits=32, format=I2S.STEREO, rate=RATE, ibuf=BUFFER_LENGTH_IN_BYTES)

samples = memoryview(bytearray(USER_BUFFER_LENGTH_IN_BYTES))

left_samples = []
right_samples = []
while True:
    num_bytes_read = i2s.readinto(samples) # read samples from the I2S peripheral
    for i in range(0, len(samples ), 8):
        left_sample = struct.unpack('<i', samples [i:i+4])[0]  
        left_samples.append(left_sample)
        right_sample = struct.unpack('<i', samples [i+4:i+8])[0]
        right_samples.append(right_sample)
        print('L: ', left_sample, '		R: ', right_sample)
`
The oscilloscope showed the right frequencies roughly, but the master clock, SCKI, generated by pwm0, is not synchronous to the WC and BCK.  The LRCK and BCK are synchronous.  (If I trigger on one of them, the other stays locked on in the same phase for every scope sweep.

Here is the LRCK (yellow) and BCK (blue).  It seems rock solid:  
![image](https://github.com/micropython/micropython/assets/994565/8acee324-0c34-4351-b3f1-d5dccec0c61d)

Here is a photo of the SCKI (yellow) and BCK (blue).  The one I don't trigger on jumps everywhere:
![image](https://github.com/micropython/micropython/assets/994565/68946d83-577d-466e-be04-23fe11d0a220)

It appears my worries were correct.  Do either of you have any suggestions?

**Reply by frohro** [reply]
I am an absolute newby on writing PIO, (never wrote any PIO at all) but it looks to me like creating the master clock I need would only entail changing the 32 bit write PIO in [machine_i2s.c](https://github.com/micropython/micropython/blob/master/ports/rp2/machine_i2s.c), as all we are doing is creating a clock that is synchronous with the others.  BCK is 1/4 as fast as the SCKI master clock we need, so it looks like we could make the base frequency of the state machine to be four times as high as it now is.  Then we need to adjust the delays and add nop instructions with the appropriate delays so everything fires at the right time with the faster clock.  The new clock signal could be added to the .side(...) statements so it is output on the next GPIO Pin.  I have my board soldered this way.  
![image](https://github.com/micropython/micropython/assets/994565/f05a9a64-1674-4969-8209-ba688ff31055)

@miketeachman, it looks like you are the author of machine_i2s.c.   Is there any chance you could help me try this out?  Am I missing something important?  It looks like we would have enough instructions (especially as I only need 32-bit receive).  It seems like it would be a good step in the direction of improving the I2S for the RP2 micropython.

**Reply by miketeachman** [reply]
Perhaps the existing [MicroPython API for the PIO](https://docs.micropython.org/en/latest/library/rp2.html) could be used to create the master clock signal on an output pin?   With a PIO I expect that it's possible to create a clock signal synchronous to the I2S signals, which were created using a PIO.   This approach would be simpler than a change to the source code.

**Reply by frohro** [reply]
I did some more work on this and discovered an anomaly in the BCK. Here is what it looks like.  
![image](https://github.com/micropython/micropython/assets/994565/11b750fc-42ff-4928-83bc-3e950aea819c)
The code is the same as I used above.  You can see where sometimes it looks like it takes about 8ns more per pulse than at other times.  This seems like a problem.  
I did some more reading, and it is because of the fractional PLL divider.  I don't know how much jitter is permissible for the PCM1808, but some is to be expected due to the fractional PLL divider.

**Reply by robert-hh** [reply]
Fractional division chimes in if the divider isn't an integer. What happens if you run the CPU at 144MHz instead of 125Mhz, using machine.freq()? or 120Mhz?

**Reply by frohro** [reply]
The jitter is still there for both frequencies.

**Comment by frohro** [comment]
That may be possible.  I like the idea that it does not require messing with the source code.  I haven't done enough with PIO to know what problems may occur, in the integration.  I am a total newby to PIO, but as I understand it, the IO pins are picked and the clock set at the same time the PIO is created.  Can you provide any direction on how to make a separate PIO program run on a contiguous pin and in synchronism with the one created by the I2S micropython instance?  Can you modify a PIO with a separate program in a way to create a new synchronous clock?  If so, how?  Thanks for the idea?

**Comment by miketeachman** [comment]
> Can you provide any direction on how to make a separate PIO program run on a contiguous pin and in synchronism with the one created by the I2S micropython instance?

There is a [blink example](https://github.com/raspberrypi/pico-micropython-examples/blob/master/pio/pio_blink.py) that would be a good starting point.

**Reply by frohro** [reply]
Here is a little test I did with PWM and PIO.  It seems there might be a difference in the frequency of the two.  I attempted to make two waves of 48kHz, one with PWM, and a second with PIO.  Maybe I have a coding mistake, but the two waves are not the same frequency and are not synchronous.  One (PIO) is 48.0008kHz and the other (pwm) is 48.0038kHz.  Here is my code: 
`import time
import rp2
from machine import Pin, I2S, PWM

machine.freq(125_000_000)
BUFFER_LENGTH_IN_BYTES = 4096
USER_BUFFER_LENGTH_IN_BYTES = round(BUFFER_LENGTH_IN_BYTES / 4)
RATE = 48000

pwm0 = PWM(Pin(3), freq=RATE, duty_u16=32768)

@rp2.asm_pio(set_init=rp2.PIO.OUT_LOW)
def blink():
    wrap_target()
    set(pins, 1)
    set(pins, 0)
    wrap()

sm = rp2.StateMachine(1, blink, freq=96000, set_base=Pin(0))
sm.active(1)
`
If the PWM and the PIO are off like this, no wonder that the pwm generated SCLK is not in sync.

**Reply by frohro** [reply]
Here is a little progress.  By using one of the PIO (only one works for this), I was able to get a master bit clock that is four times the frequency of BCK and it is synchronized with BCK.
![image](https://github.com/micropython/micropython/assets/994565/fbfad0b2-b5f1-4fb5-acda-7650ac6cf754)
There is no data, though.  Everything reads zeros.
`
from machine import Pin, I2S, PWM
import struct

BUFFER_LENGTH_IN_BYTES = 4096
USER_BUFFER_LENGTH_IN_BYTES = round(BUFFER_LENGTH_IN_BYTES / 4)
RATE = 48000

@rp2.asm_pio(set_init=rp2.PIO.OUT_LOW)
def blink():
    wrap_target()
    set(pins, 1)
    set(pins, 0)
    wrap()

sm = rp2.StateMachine(0, blink, freq=48000*256*2, set_base=Pin(3))
sm.active(1)
i2s = I2S(0, sck=Pin(0), sd=Pin(2), ws=Pin(1), mode=I2S.RX, bits=32, format=I2S.STEREO, rate=RATE, ibuf=BUFFER_LENGTH_IN_BYTES)

samples = memoryview(bytearray(USER_BUFFER_LENGTH_IN_BYTES))

left_samples = []
right_samples = []
while True:
    num_bytes_read = i2s.readinto(samples) # read samples from the I2S peripheral
    for i in range(0, len(samples ), 8):
        left_sample = struct.unpack('<i', samples [i:i+4])[0]  
        left_samples.append(left_sample)
        right_sample = struct.unpack('<i', samples [i+4:i+8])[0]
        right_samples.append(right_sample)
        print('L: ', left_sample, '		R: ', right_sample)
`

**Reply by robert-hh** [reply]
Note: to maintain code formatting, put it between lines of three backticks ```. For convenience, highlight the code and push Ctrl-E. You can as well indent the code lines by one tab to keep the formatting.

**Reply by robert-hh** [reply]
I get a similar picture with your code. In the sample code, the PIO frequency seems wrong. I had to use `freq=24_576_000`. Screen shot below. In the statistics you see in the histogram the alternating frequencies. The two clocks are synchronous once started, but not at the same phase for every start.

![rp2_I2S_PIO_3_072MHz](https://github.com/micropython/micropython/assets/12476868/bb71702a-15fe-44b2-974a-9c3357b2e511)

**Reply by frohro** [reply]
Thanks for the tip about posting code.  I wondered what was up.  I'll try it next time I try and post code.

**Reply by frohro** [reply]
@miketeachman, I don't like the jitter I am getting with the RP2 and am considering modifications of my hardware design to include a SKCI clock.  This would require making the PCM1808 master and would give me an accurate jitter free 24.576 MHz or 12.288 MHz clock which would help solve other synchronization issues as well.  The hardware cost could be quite minimal, but I am wondering about the software effort if the RP2 is no longer supplying the clocks.  I am trying to make this work with micropython, because my intended audience will benefit significantly.  Can you estimate how difficult it would be to make the software for I2S work in slave mode?
Thanks!
Rob

**Comment by miketeachman** [comment]
> Can you estimate how difficult it would be to make the software for I2S work in slave mode?

Sorry, I don't have an estimate.  It's not trivial for sure.

Is there a reason you need to use this particular ADC (or any ADC with a master clock)?  If you switched to a INMP441 device you won't have these issues as it doesn't need a master clock.

**Comment by frohro** [comment]
Thanks @miketeachman.  That is helpful.  I need a good audio ADC as I am building an SDR receiver.  In the past, [I used a USB audio card](https://fweb.wallawalla.edu/~frohro/ClassHandouts/Electronics/A%20Comparison%20of%20Affordable,%20Self-Assembled%20%20Software-Defined%20Radio%20Receivers%20Using%20Quadrature%20Sampling%20Down-Conversion.pdf) connected to the host computer.  With the ADC on the SDR board, it could be placed near the antenna, and the host could be anywhere connected via WIFI.  These days, computers don't come with stereo audio inputs.  Having the ADC on the SDR board is cheaper, and eliminates things that have to hang off the host computer, like the sound card.
